// @prisma/language-server/dist/__test__/quickFix/multi-file.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const MessageHandler_1 = require("../../lib/MessageHandler");
const MultifileHelper_1 = require("../MultifileHelper");
(0, vitest_1.test)('basic doc', async () => {
    const helper = await (0, MultifileHelper_1.getMultifileHelper)('quick-fix');
    const profile = helper.file('Profile.prisma');
    const diagnostics = (0, MessageHandler_1.handleDiagnosticsRequest)(helper.schema).get(profile.uri);
    const response = (0, MessageHandler_1.handleCodeActions)(helper.schema, profile.textDocument, {
        textDocument: {
            uri: profile.uri,
        },
        range: {
            start: profile.lineContaining('user   User').characterAfter('us'),
            end: profile.lineContaining('user   User').characterAfter('user'),
        },
        context: {
            diagnostics,
        },
    });
    const changes = response[0]?.edit?.changes;
    const updated = helper.applyChanges(changes);
    (0, vitest_1.expect)(updated).toMatchInlineSnapshot(`
    {
      "file:///quick-fix/Profile.prisma": "model Profile {
        id     String @id @default(uuid())
        userId String @unique
        user   User   @relation(fields: [userId], references: [id])
    }
    ",
    }
  `);
});
